### PR TITLE
Improve pygfx renderer

### PR DESCRIPTION
**Pull request by berendkleinhaneveld**
The pygfx renderer can now just create objects that are defined in the pygfx top package. Some attributes (`position`, `scale`, `rotation`, `matrix`) have a specific type, so the value is transformed into the expected type (Vector3, Quaternion, Matrix4).

Bonus: during the rewrite of the example, I came across a problem with the event handlers being removed/added on each mouse-hover. I found that the search for equivalency was not 'deep' enough yet, because the values within the `__closure__` can also be functions, which should also be checked for equivalency; not equality. Also added a test for this scenario.

### CODE REVIEW

File: collagraph/__init__.py [hunk 1]
```diff
@@ -583,7 +583,12 @@ def equivalent_code(a, b):
     ]:
         attr_a = getattr(a, attr, None)
         attr_b = getattr(b, attr, None)
-        if attr_a != attr_b:
+        if attr in ["co_freevars", "co_varnames"]:
```
Comment (Korijn) [inline collagraph/__init__.py:586]: just a stylistic thing, but in these situations I would recommend using sets, containment test is O(1)

File: collagraph/__init__.py [hunk 2]
```diff
@@ -583,7 +583,12 @@ def equivalent_code(a, b):
...
+            # co_varnames and co_freevars can contain names, which might
+            # be different but should be similar at least in length
+            if len(attr_a) != len(attr_b):
```
Comment (Korijn) [inline collagraph/__init__.py:589]: I don't really understand why using `len()` is sufficient here?

File: collagraph/__init__.py [hunk 3]
```diff
@@ -560,14 +560,14 @@ def equivalent_code(a, b):
     In order to determine this, a number of properties are compared that
     should be equal between similar functions.
     It checks all co_* props of __code__ (as seen in Python 3.9), except for:
-      * co_varnames
       * co_firstlineno
       * co_lnotab
       * co_name
     because those vars can differ without having any impact on the equivalency
     of the functions themselves.
     """
```
Comment (Korijn) [inline collagraph/__init__.py:568]: I would recommend moving `equivalent_functions` and `equivalent_code` to a separate module, it's a pretty complicated beast and it's not related to the class it is in right now.

File: collagraph/renderers/pygfx_renderer.py [hunk 4]
```diff
@@ -39,42 +25,21 @@ def insert(
...
-            obj.material.color = value
-            return
-
-        if hasattr(value, "__len__") and len(value) == 3:
+        if attr == "position" or attr == "scale":
             value = gfx.linalg.Vector3(*value)
+        elif attr == "rotation":
+            value = gfx.linalg.Quaternion(*value)
+        elif attr == "matrix":
+            value = gfx.linalg.Matrix4(*value)
```
Comment (Korijn) [inline collagraph/renderers/pygfx_renderer.py:34]: I'm expecting this to break down at some point, but if it works for now, sure

File: examples/pygfx-example.py [hunk 5]
```diff
@@ -92,7 +77,7 @@ def random_point(index, selected, hovered):
                 ],
                 "key": index,
                 "onClick": lambda event: set_selected(index),
-                "onHover": lambda event: set_hovered(index),
+                "onPointer_move": lambda event: set_hovered(index),
```
Comment (Korijn) [inline examples/pygfx-example.py:80]: what is up with this casing?

File: examples/pygfx-example.py [hunk 6]
```diff
@@ -168,34 +119,16 @@ def toggle():
...
-                "position": [4, 4, -4],
-            },
-        ),
-        h(
-            "Point",
+            "Mesh",
             {
                 "name": "Hip",
                 "position": [2, 2, 2],
                 "color": [1, 0.4, 0],
```
Comment (Korijn) [inline examples/pygfx-example.py:129]: color on Mesh directly? what does that do?

File: collagraph/__init__.py (hunk 2 above)
Comment (berendkleinhaneveld) [reply collagraph/__init__.py:589]: The contents of the `co_freevars` and `co_varnames` contain names for variables that are used within the function. A function that is similar in logic but has different variable names is still equivalent, so we can't check for strict equality. If the length is not the same, then we already know that the logic of the functions can't be equal, so we can bail.

File: collagraph/renderers/pygfx_renderer.py (hunk 4 above)
Comment (berendkleinhaneveld) [reply collagraph/renderers/pygfx_renderer.py:34]: Yes, this is highly dependent on the pygfx API, so we should make sure to keep this in sync.

File: examples/pygfx-example.py (hunk 6 above)
Comment (berendkleinhaneveld) [reply examples/pygfx-example.py:129]: I think this is still some leftover code :) I'll remove it!

File: collagraph/__init__.py (hunk 2 above)
Comment (Korijn) [reply collagraph/__init__.py:589]: Ah, so it's just an early exit?

File: examples/pygfx-example.py (hunk 5 above)
Comment (berendkleinhaneveld) [reply examples/pygfx-example.py:80]: All the cAsInG_s!

File: collagraph/__init__.py (hunk 1 above)
Comment (Korijn) [reply collagraph/__init__.py:586]:
```suggestion
        if attr in {"co_freevars", "co_varnames"}:
```

File: collagraph/renderers/pygfx_renderer.py (hunk 4 above)
Comment (Korijn) [inline collagraph/renderers/pygfx_renderer.py:34]:
```suggestion
        if isinstance(obj, gfx.WorldObject):
            if attr == "position" or attr == "scale":
                value = gfx.linalg.Vector3(*value)
            elif attr == "rotation":
                value = gfx.linalg.Quaternion(*value)
            elif attr == "matrix":
                value = gfx.linalg.Matrix4(*value)
```

Comment (Korijn) [reply collagraph/renderers/pygfx_renderer.py:34]: I put a suggestion in a separate comment to explain better what I mean

Comment (berendkleinhaneveld) [reply collagraph/renderers/pygfx_renderer.py]: Ah, I might not have seen your edited comment :) It's clear now though: great suggestion!

File: collagraph/__init__.py (hunk 2 above)
Comment (berendkleinhaneveld) [reply collagraph/__init__.py:589]: Exactly!

File: examples/pygfx-example.py (hunk 5 above)
Comment (berendkleinhaneveld) [reply examples/pygfx-example.py:80]: Yes, let's go for pythonic! 🐍

**Pull request merged**